openfaas/faas-netes
# Function Profiles

## Description

- Create new Policy client and type so that we can read Policy
  configurations from the k8s cluster.  We assume that Policy ConfigMaps
  live in the same namespace as the function deployment. This
  implementation allows the admin to define tolerations that can
  be applied via the policy annoation.

## Motivation and Context



Relates to #586 

## How Has This Been Tested?




I have done one manual test as follows:

1. create a cluster and deploy the test image

```sh
kind create cluster

arkade install openfaas --clusterrole --basic-auth=false --set faasnetes.image=theaxer/faas-netes:policies

kubectl port-forward -n openfaas svc/gateway 8080:8080 > /dev/null 2>&1 &
```

2. edit the cluster role to enable access to ConfigMaps

```sh
kubectl edit clusterrole openfaas-controller
```

3. create this stack file
```yaml
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080

functions:
  stronghash:
    skip_build: true
    image: functions/alpine:latest
    fprocess: "sha512sum"
    annotations:
        com.openfaas.policies: test
```

4.  try to deploy the function and see an error

```sh
faas-cli deploy -f stack.yml
```

5. Create this configmap file
```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: test
  namespace: openfaas-fn
data:
  # Configuration values can be set as key-value properties
  policy: |
    tolerations:
    - key: "example-key"
      operator: "Exists"
      effect: "NoSchedule"

```

6.  try to deploy the function successfully

```sh
faas-cli deploy -f stack.yml
```

7. describe the function and note the tolerations section, it should contain `example-key:NoSchedule`

```sh
kubectl -n openfaas-fn describe po stronghash-788786d548-mdq7n
```


## Types of changes


## Checklist:
